test/helpers/foundry/extract.ts

```diff
@@ -0,0 +1,211 @@
+          );
+        }
+      }
+      // add the *final* path to the list of binaries
+      paths.push(join(dir, relative(tempDir, path)));
+    }
+
+    // everything has been extracted; move the files to their final destination
+    await rename(tempDir, dir);
+    // return the list of extracted binaries
+    return paths;
+  } catch (e) {
+    // if things fail for any reason try to clean up a bit. it is very important
+    // to not leave `dir` behind, as its existence is a signal that the binaries
+    // are installed.
+    await Promise.all([rm(tempDir, rmOpts), rm(dir, rmOpts)]).catch(noop);
+    throw e;
+  }
+}
+/**
+ * Extracts the binaries from a tar archive.
+ *
+ * @param url - The URL of the archive to extract the binaries from
+ * @param binaries - The list of binaries to extract
+ * @param dir - The destination directory
+ * @param checksumAlgorithm - The checksum algorithm to use
+ * @returns The list of binaries extracted
+ */
+
+async function extractFromTar(
+  url: URL,
+  binaries: string[],
+  dir: string,
+  checksumAlgorithm?: string,
+) {
+  const downloads: {
+    path: string;
+    binary: string;
+    checksum?: string;
+  }[] = [];
+  await pipeline(
+    startDownload(url),
+    extractTar(
+      {
+        cwd: dir,
+        transform: (entry) => {
+          const absolutePath = entry.absolute;
+          if (!absolutePath) {
+            throw new Error('Missing absolute path for entry');
+          }
+
+          if (checksumAlgorithm) {
+            const hash = createHash(checksumAlgorithm);
+            const passThrough = new Minipass({ async: true });
+            passThrough.pipe(hash);
+            passThrough.on('end', () => {
+              downloads.push({
+                path: absolutePath,
+                binary: entry.path,
+                checksum: hash.digest('hex'),
+              });
+            });
+            return passThrough;
+          }
+
+          // When no checksum is needed, record the entry and return undefined
+          // to use the original stream without transformation
+          downloads.push({
+            path: absolutePath,
+            binary: entry.path,
+          });
+          return undefined;
+        },
+      },
+      binaries,
+    ),
+  );
+  return downloads;
+}
+/**
+ * Extracts the binaries from a zip archive.
+ *
+ * @param url - The URL of the archive to extract the binaries from
+ * @param binaries - The list of binaries to extract
+ * @param dir - The destination directory
+ * @param checksumAlgorithm - The checksum algorithm to use
+ * @returns The list of binaries extracted
+ */
+
+async function extractFromZip(
+  url: URL,
+  binaries: string[],
+  dir: string,
+  checksumAlgorithm?: string,
+) {
+  const agent = new (url.protocol === 'http:' ? HttpAgent : HttpsAgent)({
+    keepAlive: true,
+  });
+  const source: Source = {
+    async size() {
+      const download = startDownload(url, { agent, method: 'HEAD' });
```

Comment:
We just request the headers from github, we want to know how big the zip file is so the `unzipper` package can ask for the zip's "central directory" (an index of all the files in the zip), which is always near the end of the zip file.

The `download` variable here is a `PassThrough` stream, which, for us, is just a stream that acts a middleman between the source stream and a destination stream.

I'll explain a bit more in the `startDownload` function below.